README.md
# @rappopo/dab-couch

A [Rappopo DAB](https://github.com/rappopo/dab) implementation for CouchDB 2.0 and above.

## Installation

Simply invoke this command in your project folder:

```
$ npm install --save @rappopo/dab-couch
```

And within your script:

```javascript
const DabCouch = require('@rappopo/dab-couch')
const dab = new DabCouch({
  url: 'http://localhost:5984',
  dbName: 'mydb'
})
...
dab.findOne('my-doc').then(function(doc) { ... })
```

## Options

`url`: your CouchDB url endpoint. If it not provided, it defauts to: *http://localhost:5984*

`dbName`: the database name to connect to. Defaults to *test*

`retainOnRemove`: array of columns to retain when a document is deleted. Default: []. 

When CouchDB delete a document, it actually PUTs a document with content like this:

```javascript
{
  "_id": "<doc_id>",
  "_rev": "<rev_id>",
  "_deleted": true
}
```

But sometimes you want to also have some columns to be put on that deleted document. The `retainOnRemove` simply left those columns intact, e.g:

```javascript
{
  "_id": "<doc_id>",
  "_rev": "<rev_id>",   
  "_deleted": true,
  "type": "ADDRESS"
}
```

## Features

* [x] [find](https://docs.rappopo.com/dab/method/find/)
* [x] [findOne](https://docs.rappopo.com/dab/method/find-one/)
* [x] [create](https://docs.rappopo.com/dab/method/create/)
* [x] [update](https://docs.rappopo.com/dab/method/update/)
* [x] [remove](https://docs.rappopo.com/dab/method/remove/)
* [x] [bulkCreate](https://docs.rappopo.com/dab/method/bulk-create/)
* [x] [bulkUpdate](https://docs.rappopo.com/dab/method/bulk-update/)
* [x] [bulkRemove](https://docs.rappopo.com/dab/method/bulk-remove/)
* [x] [copyFrom](https://docs.rappopo.com/dab/method/copy-from/)
* [x] [copyTo](https://docs.rappopo.com/dab/method/copy-to/)

## Misc

* [ChangeLog](CHANGELOG.md)
* Donation: Bitcoin **16HVCkdaNMvw3YdBYGHbtt3K5bmpRmH74Y**

## License

(The MIT License)

Copyright © 2017 Ardhi Lukianto <ardhi@lukianto.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
# Changelog

## 0.6.7

* Updated to @rappopo/dab@0.6.13

## 0.6.6

* Updated to @rappopo/dab@0.6.12

## 0.6.5

* Remove debug code

## 0.6.4

* Updated to @rappopo/dab@0.6.11

## 0.6.2

* Updated to @rappopo/dab@0.6.10

## 0.6.1

* Adding rebuild to createCollection
* Adding drop to removeCollection

## 0.6.0

* Updated to @rappopo/dab@0.6.9
* Standard JS compliance

## 0.0.9

* Updated to @rappopo/dab@0.0.11

## 0.0.8

* Updated to @rappopo/dab@0.0.8

## 0.0.7

* Updated to @rappopo/dab@0.0.7
* Removing idSrc & idDest in favor of schema mask provided by @rappopo/dab

## 0.0.6

* withDetail params in all bulk methods
* Updated to @rappopo/dab@0.0.6 to have copyFrom & copyTo for free

## 0.0.5

* Spec changed on bulkRemove

## 0.0.4

* Adding .asCallback to the Promise

## 0.0.3

* Update README

## 0.0.2

* Adding bulkCreate method
* Adding bulkUpdate method
* Adding bulkRemove method

## 0.0.1

* First commit ever
* Support the most basic methods:
  * find
  * findOne
  * create
  * update
  * remove

test/_lib.js
'use strict'

const _ = require('lodash')
const async = require('async')

module.exports = {
  _: _,
  options: {
    url: 'http://admin:admin@localhost:5984'
  },
  schema: {
    name: 'test'
  },
  schemaDummy: {
    name: 'test1'
  },
  schemaFull: {
    name: 'full',
    attributes: {
      _id: 'string',
      name: 'string',
      age: 'integer'
    },
    indexes: {
      name: true,
      age: true
    }
  },
  schemaHidden: {
    name: 'hidden',
    attributes: {
      _id: 'string',
      name: { type: 'string', hidden: true },
      age: 'integer'
    }
  },
  schemaMask: {
    name: 'mask',
    attributes: {
      _id: { type: 'string', mask: 'id' },
      name: { type: 'string', mask: 'fullname' },
      age: { type: 'integer' }
    }
  },
  schemaMaskDummy: {
    name: 'mask1',
    attributes: {
      _id: { type: 'string', mask: 'id' },
      name: { type: 'string', mask: 'fullname' },
      age: { type: 'integer' }
    }
  },
  schemaBulk: {
    name: 'test1'
  },
  docs: [
    { _id: 'jack-bauer', name: 'Jack Bauer' },
    { _id: 'johnny-english', name: 'Johnny English' },
    { name: 'Jane Boo', age: 20 }
  ],
  docsMask: [
    { id: 'jack-bauer', fullname: 'Jack Bauer' },
    { id: 'johnny-english', fullname: 'Johnny English' },
    { fullname: 'Jane Boo', age: 20 }
  ],
  timeout: 5000,
  resetDb: function (callback, fillIn = true) {
    let me = this
    const nano = require('nano')(me.options.url)
    async.mapSeries(['schema', 'schemaFull', 'schemaHidden', 'schemaMask', 'schemaBulk',
      'schemaDummy', 'schemaMaskDummy'], function (s, callb) {
      nano.db.destroy(me[s].name, function (err, body) {
        if (err && err.statusCode !== 404) return callb(err)
        nano.db.create(me[s].name, function (err, body) {
          if (err) return callb(err)
          let db = nano.db.use(me[s].name)
          if (['test1', 'mask1'].indexOf(me[s].name) > -1 || !fillIn) return callb(null, null)
          db.bulk({ docs: me.docs }, function (err, results) {
            if (err) return callb(err)
            async.mapSeries(['name', 'age'], function (i, cb) {
              db.createIndex({
                index: { fields: [i] },
                name: i
              }, function (e) {
                cb(null)
              })
            }, function () {
              callb()
            })
          })
        })
      })
    }, callback)
  }
}


test/01-find.js
'use strict'

const chai = require('chai')
const chaiSubset = require('chai-subset')
const expect = chai.expect

chai.use(chaiSubset)

const Cls = require('../index')
const lib = require('./_lib')

describe('find', function () {
  before(function (done) {
    this.timeout(lib.timeout)
    lib.resetDb(function (err) {
      if (err) throw err
      done()
    })
  })

  it('should return error if collection doesn\'t exist', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'none' })
      })
      .catch(err => {
        expect(err).to.be.a('error').and.have.property('message', 'Collection not found')
        done()
      })
  })

  it('should return empty value', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'test', query: { _id: 'no-agent' } })
      })
      .then(result => {
        expect(result.data).to.be.a('array').and.have.length(0)
        done()
      })
  })

  it('should return all values', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'test' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        expect(result.data).to.be.a('array').and.containSubset(lib.docs)
        done()
      })
  })

  it('should return filtered values', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'test', query: { _id: 'jack-bauer' } })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(1)
        expect(result.data).to.be.a('array').and.containSubset([lib.docs[0]])
        done()
      })
  })

  it('should return 2nd page', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'test', page: 2, limit: 1, sort: [{ name: 'asc' }] })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        expect(result.data).to.be.a('array').that.have.length(1)
        expect(result.data[0]).to.include(lib.docs[2])
        done()
      })
  })

  it('should sort in descending order', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.find({ collection: 'test', sort: [{ name: 'desc' }] })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        let keys = lib._.map(result.data, 'name')
        expect(keys).to.eql(['Johnny English', 'Jane Boo', 'Jack Bauer'])
        done()
      })
  })

  it('should return enforced values according to its definitions', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaFull)
      .then(result => {
        return cls.find({ collection: 'full' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        expect(result.data).to.containSubset([{ _id: 'jack-bauer', name: 'Jack Bauer', age: null }])
        expect(result.data).to.containSubset([{ _id: 'johnny-english', name: 'Johnny English', age: null }])
        expect(result.data).to.containSubset([{ name: 'Jane Boo', age: 20 }])
        done()
      })
  })

  it('should return enforced values with hidden columns', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaHidden)
      .then(result => {
        return cls.find({ collection: 'hidden' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        expect(result.data).to.containSubset([{ _id: 'jack-bauer', age: null }])
        expect(result.data).to.containSubset([{ _id: 'johnny-english', age: null }])
        expect(result.data).to.containSubset([{ age: 20 }])
        done()
      })
  })

  it('should return enforced values with masks', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaMask)
      .then(result => {
        return cls.find({ collection: 'mask' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        // expect(result.total).to.equal(3)
        expect(result.data).to.containSubset([{ id: 'jack-bauer', fullname: 'Jack Bauer', age: null }])
        expect(result.data).to.containSubset([{ id: 'johnny-english', fullname: 'Johnny English', age: null }])
        expect(result.data).to.containSubset([{ fullname: 'Jane Boo', age: 20 }])
        done()
      })
  })
})


test/03-create.js
'use strict'

const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')
const expect = chai.expect

chai.use(chaiAsPromised)

const Cls = require('../index')
const lib = require('./_lib')
const body = {
  _id: 'jason-bourne',
  name: 'Jason Bourne'
}

describe('create', function () {
  beforeEach(function (done) {
    this.timeout(lib.timeout)
    lib.resetDb(function (err) {
      if (err) throw err
      done()
    })
  })

  it('should return error if collection doesn\'t exist', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.create(body, { collection: 'none' })
      })
      .catch(err => {
        expect(err).to.be.a('error').and.have.property('message', 'Collection not found')
        done()
      })
  })

  it('should return error if doc exists', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.create(lib.docs[0], { collection: 'test' })
      })
      .catch(err => {
        expect(err).to.be.a('error').and.have.property('message', 'Document already exists')
        done()
      })
  })

  it('should return the correct value', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schema)
      .then(result => {
        return cls.create(body, { collection: 'test' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        expect(result.data).to.have.property('_id', 'jason-bourne')
        expect(result.data).to.have.property('name', 'Jason Bourne')
        done()
      })
  })

  it('should return enforced values according to its definitions', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaFull)
      .then(result => {
        return cls.create(body, { collection: 'full' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        expect(result.data).to.have.property('_id', 'jason-bourne')
        expect(result.data).to.have.property('name', 'Jason Bourne')
        expect(result.data).to.have.property('age', null)
        done()
      })
  })

  it('should return enforced values with hidden columns', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaHidden)
      .then(result => {
        return cls.create(body, { collection: 'hidden' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        expect(result.data).to.have.property('_id', 'jason-bourne')
        expect(result.data).to.not.have.property('name')
        expect(result.data).to.have.property('age', null)
        done()
      })
  })

  it('should return enforced values with masks', function (done) {
    const cls = new Cls(lib.options)
    cls.createCollection(lib.schemaMask)
      .then(result => {
        return cls.create({ id: 'jason-bourne', fullname: 'Jason Bourne' }, { collection: 'mask' })
      })
      .then(result => {
        expect(result.success).to.equal(true)
        expect(result.data).to.have.property('id', 'jason-bourne')
        expect(result.data).to.have.property('fullname', 'Jason Bourne')
        expect(result.data).to.have.property('age', null)
        done()
      })
  })
})


index.js
'use strict'

const nano = require('nano')
const Dab = require('@rappopo/dab').Dab

class DabCouch extends Dab {
  constructor (options) {
    super(options)
    this.client = {}
  }

  setOptions (options) {
    super.setOptions(this._.merge(this.options, {
      url: options.url || 'http://localhost:5984',
      retainOnRemove: options.retainOnRemove || []
    }))
    this.nano = nano(this.options.url)
  }

  setClient (params) {
    params = params || {}
    if (this.client[params.collection]) return
    if (this._.keys(this.collection).indexOf(params.collection) === -1) return new Error('Collection not found')
    this.client[params.collection] = this.nano.use(params.collection)
  }

  _rebuildSchema (coll) {
    return new Promise((resolve, reject) => {
      let rebuild = !this._.isEmpty(this.collection[coll].attributes)
      if (!rebuild) return resolve(true)
      this._dropDb(coll)
        .then(result => {
          return this.nano.db.create(coll)
        })
        .then(result => {
          return this._rebuildIndex(coll)
        })
        .then(result => {
          resolve(true)
        })
        .catch(reject)
    })
  }

  _rebuildIndex (coll) {
    return new Promise((resolve, reject) => {
      let rebuild = !this._.isEmpty(this.collection[coll].indexes)
      if (!rebuild) return resolve(true)
      let indexes
      this.nano.db.get(coll)
        .then(result => {
          return this.nano.request({
            db: coll,
            doc: '_index',
            method: 'get'
          })
        })
        .then(result => {
          indexes = this._.filter(result.indexes, i => {
            return i.name !== '_all_docs'
          })
          if (indexes.length === 0) return true
          let proms = []
          this._.each(indexes, i => {
            proms.push(this.nano.request({
              db: coll,
              path: '/_index/' + i.ddoc + '/json/' + i.name,
              method: 'delete'
            }))
          })
          return Promise.all(proms)
        })
        .then(result => {
          let proms = []
          this._.forOwn(this.collection[coll].indexes, (v, k) => {
            proms.push(this.nano.use(coll).createIndex({
              name: k,
              index: { fields: v.column }
            }))
          })
          return Promise.all(proms)
        })
        .then(result => {
          resolve(true)
        })
        .catch(reject)
    })
  }

  _dropDb (coll) {
    return new Promise((resolve, reject) => {
      this.nano.db.get(coll, (e, i) => {
        if (e && e.statusCode !== 404) return reject(e)
        Promise.resolve(true)
          .then(_ => {
            if (i) return this.nano.db.destroy(coll)
            return true
          })
          .then(result => {
            resolve(true)
          })
          .catch(reject)
      })
    })
  }

  createCollection (coll, params) {
    params = params || {}
    return new Promise((resolve, reject) => {
      super.createCollection(coll)
        .then(result => {
          if (params.rebuild) return this._rebuildSchema(coll.name)
          return true
        })
        .then(result => {
          let e = this.setClient({ collection: coll.name })
          if (e instanceof Error) return reject(e)
          resolve(result)
        })
        .catch(reject)
    })
  }

  renameCollection (oldName, newName, params) {
    params = params || {}
    return Promise.reject(new Error('Not implemented'))
  }

  removeCollection (name, params) {
    params = params || {}
    let drop = params.drop && this.collection[name]
    return new Promise((resolve, reject) => {
      super.removeCollection(name)
        .then(result => {
          if (!drop) return true
          return this._dropDb(name)
        })
        .then(result => {
          delete this.client[name]
          resolve(true)
        })
        .catch(reject)
    })
  }

  find (params) {
    [params] = this.sanitize(params)
    let limit = params.limit || this.options.limit
    let skip = ((params.page || 1) - 1) * limit
    let query = params.query || {}

    let sortKeys = this._.keys(params.sort) || []
    if (sortKeys.length > 0) {
      let qidx = {}
      this._.each(sortKeys, k => {
        qidx[k] = { $gte: null }
      })
      query = this._.merge(query, qidx)
    }

    return new Promise((resolve, reject) => {
      let e = this.setClient(params)
      if (e instanceof Error) return reject(e)
      let q = {
        selector: query,
        limit: limit,
        skip: skip,
        execution_stats: true
      }
      if (params.sort) {
        let sort = []
        this._.forOwn(params.sort, (v, k) => {
          let o = {}
          o[k] = v === -1 ? 'desc' : 'asc'
          sort.push(o)
        })
        q.sort = sort
      }
      this.client[params.collection].find(q)
        .then(result => {
          let data = { success: true, data: [] }
          result.docs.forEach((d, i) => {
            data.data.push(this.convert(d, { collection: params.collection }))
          })
          resolve(data)
        })
        .catch(reject)
    })
  }

  _findOne (id, params, callback) {
    let e = this.setClient(params)
    if (e instanceof Error) {
      return callback(null, {
        success: false,
        err: e
      })
    }
    this.client[params.collection].get(id, params.options || {}, (err, result) => {
      if (err) {
        return callback(null, {
          success: false,
          err: err.statusCode === 404 ? new Error('Document not found') : err
        })
      }
      callback(null, {
        success: true,
        data: result
      })
    })
  }

  findOne (id, params) {
    [params] = this.sanitize(params)
    this.setClient(params)
    return new Promise((resolve, reject) => {
      this._findOne(id, params, (e, result) => {
        if (!result.success) return reject(result.err)
        let data = {
          success: true,
          data: this.convert(result.data, { collection: params.collection })
        }
        resolve(data)
      })
    })
  }

  _create (body, params, callback) {
    let e = this.setClient(params)
    if (e instanceof Error) {
      return callback(null, {
        success: false,
        err: e
      })
    }
    this.client[params.collection].insert(body, params.options || {}, (err, result) => {
      if (err) {
        return callback(null, {
          success: false,
          err: err
        })
      }
      this._findOne(result.id, params, callback)
    })
  }

  create (body, params) {
    [params, body] = this.sanitize(params, body)
    return new Promise((resolve, reject) => {
      if (body._id) {
        this._findOne(body._id, params, (e, result) => {
          if (result.success) return reject(new Error('Document already exists'))
          this._create(body, params, (e, result) => {
            if (!result.success) return reject(result.err)
            result.data = this.convert(result.data, { collection: params.collection })
            resolve(result)
          })
        })
      } else {
        this._create(body, params, (e, result) => {
          if (!result.success) return reject(result.err)
          result.data = this.convert(result.data, { collection: params.collection })
          resolve(result)
        })
      }
    })
  }

  update (id, body, params) {
    [params, body] = this.sanitize(params, body)
    body = this._.omit(body, ['_id'])
    return new Promise((resolve, reject) => {
      this._findOne(id, params, (e, result) => {
        if (!result.success) return reject(result.err)
        let source = result.data
        if (params.fullReplace) {
          body._id = id
          body._rev = result.data._rev
        } else {
          body = this._.merge(result.data, body)
        }
        this._create(body, params, (e, result) => {
          if (!result.success) return reject(result.err)
          result.data = this.convert(result.data, { collection: params.collection })
          if (params.withSource) result.source = this.convert(source, { collection: params.collection })
          resolve(result)
        })
      })
    })
  }

  remove (id, params) {
    [params] = this.sanitize(params)
    this.setClient(params)
    return new Promise((resolve, reject) => {
      let e = this.setClient(params)
      if (e instanceof Error) return reject(e)
      this._findOne(id, params, (e, result) => {
        if (!result.success) return reject(result.err)
        let source = result.data
        let newBody = {
          _id: id,
          _rev: source._rev,
          _deleted: true
        }
        this._.each(this.options.retainOnRemove, r => {
          if (this._.has(source, r)) newBody[r] = source[r]
        })
        this.client[params.collection].insert(newBody, params.options || {}, (err, result) => {
          if (err) return reject(result.err)
          let data = {
            success: true
          }
          if (params.withSource) data.source = this.convert(source, { collection: params.collection })
          resolve(data)
        })
      })
    })
  }

  bulkCreate (body, params) {
    [params, body] = this.sanitize(params, body)
    return new Promise((resolve, reject) => {
      let e = this.setClient(params)
      if (e instanceof Error) return reject(e)
      if (!this._.isArray(body)) return reject(new Error('Requires an array'))
      this._.each(body, (b, i) => {
        if (!b._id) b._id = this.uuid()
        body[i] = this._.omit(b, ['_rev', '_deleted'])
      })
      const keys = this._(body).map('_id').value()
      this.client[params.collection].fetch({
        keys: keys
      }, (err, result) => {
        if (err) return reject(err)
        let info = result.rows
        this.client[params.collection].bulk({ docs: body }, (err, result) => {
          if (err) return reject(err)
          let ok = 0
          let status = []
          this._.each(result, (r, i) => {
            let stat = { success: Boolean(r.ok) }
            stat._id = r.id
            if (!stat.success) stat.message = info[i] && info[i].value ? 'Document already exists' : this._.upperFirst(r.name)
            else ok++
            status.push(stat)
          })
          let data = {
            success: true,
            stat: {
              ok: ok,
              fail: body.length - ok,
              total: body.length
            }
          }
          if (params.withDetail) data.detail = status
          resolve(data)
        })
      })
    })
  }

  bulkUpdate (body, params) {
    [params, body] = this.sanitize(params, body)
    this.setClient(params)
    return new Promise((resolve, reject) => {
      let e = this.setClient(params)
      if (e instanceof Error) return reject(e)
      if (!this._.isArray(body)) return reject(new Error('Requires an array'))
      this._.each(body, (b, i) => {
        if (!b._id) b._id = this.uuid() // will likely to introduce 'not-found'
        body[i] = this._.omit(b, ['_rev', '_deleted'])
      })
      const keys = this._(body).map('_id').value()
      this.client[params.collection].fetch({
        keys: keys
      }, (err, result) => {
        if (err) return reject(err)
        let info = result.rows
        // add rev for known doc
        this._.each(body, (b, i) => {
          if (info[i] && info[i].value) body[i]._rev = info[i].value.rev
          else body[i]._rev = '1-' + this.uuid() // will introduce purposed conflict
        })
        this.client[params.collection].bulk({ docs: body }, (err, result) => {
          if (err) return reject(err)
          let ok = 0
          let status = []
          this._.each(result, (r, i) => {
            let stat = { success: Boolean(r.ok) }
            stat._id = r.id
            if (!stat.success) stat.message = info[i] && info[i].error === 'not_found' ? 'Document not found' : this._.upperFirst(r.name)
            else ok++
            status.push(stat)
          })
          let data = {
            success: true,
            stat: {
              ok: ok,
              fail: body.length - ok,
              total: body.length
            }
          }
          if (params.withDetail) data.detail = status
          resolve(data)
        })
      })
    })
  }

  bulkRemove (body, params) {
    [params, body] = this.sanitize(params, body)
    this.setClient(params)
    return new Promise((resolve, reject) => {
      let e = this.setClient(params)
      if (e instanceof Error) return reject(e)
      if (!this._.isArray(body)) return reject(new Error('Requires an array'))
      this._.each(body, (b, i) => {
        body[i] = b || this.uuid()
      })
      this.client[params.collection].fetch({
        keys: body
      }, (err, result) => {
        if (err) return reject(err)
        let info = result.rows
        // add rev for known doc
        this._.each(body, (b, i) => {
          let newB = {
            _deleted: true,
            _id: b
          }
          if (info[i] && info[i].value) {
            newB._rev = info[i].value.rev
            newB = this._.merge(newB, this._.pick(info[i].value.doc, this.options.retainOnRemove))
          } else {
            newB._rev = '1-' + this.uuid() // will introduce purposed conflict
          }
          body[i] = newB
        })
        this.client[params.collection].bulk({ docs: body }, (err, result) => {
          if (err) return reject(err)
          let ok = 0
          let status = []
          this._.each(result, (r, i) => {
            let stat = { success: Boolean(r.ok) }
            stat._id = r.id
            if (!stat.success) stat.message = info[i] && info[i].error === 'not_found' ? 'Document not found' : this._.upperFirst(r.name)
            else ok++
            status.push(stat)
          })
          let data = {
            success: true,
            stat: {
              ok: ok,
              fail: body.length - ok,
              total: body.length
            }
          }
          if (params.withDetail) data.detail = status
          resolve(data)
        })
      })
    })
  }
}

module.exports = DabCouch
